server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -179,6 +179,13 @@ private boolean shouldSync(boolean didRefresh, boolean skipPrimaryTermCheck) {
             // Below check ensures that if there is commit, then that gets picked up by both 1st and 2nd shouldSync call.
             || isRefreshAfterCommitSafe()
             || isRemoteSegmentStoreInSync() == false;
+        // Ignore syncing segments if the underlying shard is closed
+        // This also makes sure that retries are not scheduled for shards
+        // with failed syncSegments invocation after they are closed
+        if (shardClosed()) {
+            logger.info("Shard is already closed, will stop scheduling retries");
+            return false;
+        }
```

Comment:
Should we move this check at the start of the method? Also, lets add an assert for checking that engine is closed whenever we are returning from the if shard closed block.

Reply:
We cannot add check for engineRef == null as there can be a possible race condition wherein the IndexShard is marked as CLOSED but the exact subsequent line to change the engineRef has not run under `IndexShard#close`

For now, I think it would be safe to assume that if an IndexShard is marked as `CLOSED` the engine would also be closed and stop syncing segments to remote from that point onwards.